Enable Tailscale's builtin inbound HTTPS proxy

lmagyar opened the pull request:
Replaces PR #89 #134

# Proposed Changes

This is a better solution to the #62 issue than #89 #134 because it doesn't need any manual configuration and also provides a built-in proxy for the https traffic. We don't need to refresh the certificates manually, this is all done by tailscale automatically.

## Related Issues

Fixes #62

lmagyar commented:
not stale

lmagyar commented:
I force pushed a rebase onto the lates PR #133, because with tailscale 1.38.1 some script changes are required. So this PR works only after that is merged.

frenck commented on LICENSE.md:
```diff
@@ -1,6 +1,6 @@
 # MIT License
 
-Copyright (c) 2021-2022 Franck Nijhof
+Copyright (c) 2021-2023 Franck Nijhof
```
PR is full of unrelated changed.

frenck reviewed (changes requested):
The PR references a lot, but it doesn't describe what it does.

lmagyar commented:
I stashed and rebased it.

This is an optional functionality, if the user doesn't configure anything, it works as previously, but emits some info log messages, that it can be configured and used.

If configured, this solves the unsecure connection problem (I know, it's secure, only the browsers have problem with it): downloads and regularly refreshes a tailscale certificate for our tailnet and fires up it's own built-in proxy.

lmagyar commented:
I tried to keep this add-on as zero-config as it was, all the configuration is done on the tailscale admin web UI.

frenck commented on tailscale/DOCS.md:
```diff
@@ -17,8 +17,55 @@ the following URL:
 
-You can also create an account during the add-on installation processes,
-however, it is nice to know where you need to go later on.
+<details>
+<summary>
```
Let's not do those weird collapsable, this is not a 10 pages doc.
The fact you have to write `(click to open)` is already a good indication it is bad ux.

frenck commented on tailscale/DOCS.md:
```diff
@@ -17,8 +17,55 @@ the following URL:
+
+### Home Assistant configuration for Tailscale Proxy feature
+
+You must configure Home Assistant to **not** use SSL certificates, to be
+accessible through plain http connection. The Tailscale https Proxy will access
+Home Assistant through `localhost` and will not accept a real certificate,
+connection will be closed with `proxy error: x509: cannot validate certificate
+for 127.0.0.1 because it doesn't contain any IP SANs`
```
This is the default, I think just mention that Home Assistant itself should not use certificates and link to the official documentation is enough.

frenck commented on tailscale/DOCS.md:
```diff
@@ -17,8 +17,55 @@ the following URL:
+connection will be closed with `proxy error: x509: cannot validate certificate
+for 127.0.0.1 because it doesn't contain any IP SANs`
+
+If you still want to use another https connection to access Home Assistant
+through another network, please use the **NGINX Home Assistant SSL proxy**
+add-on.
```
Just stating to use a reverse proxy should be enough. I generally like to withhold recommendations of other add-ons.

frenck commented on tailscale/DOCS.md:
````diff
@@ -17,8 +17,55 @@ the following URL:
+through another network, please use the **NGINX Home Assistant SSL proxy**
+add-on.
+
+So these lines have to be removed from your `/config/configuration.yaml`:
+
+```yaml
+http:
+#  ssl_certificate: /ssl/fullchain.pem
+#  ssl_key: /ssl/privkey.pem
+```
````
These are not there by default, the example is even unneeded, as `http` is a default loaded component.

This section can be removed.

frenck commented on tailscale/DOCS.md:
````diff
@@ -17,8 +17,55 @@ the following URL:
+#  ssl_key: /ssl/privkey.pem
+```
+
+Since Home Assistant by default blocks requests from proxies/reverse proxies,
+you need to tell your instance to allow requests from the Tailscale add-on.
+In order to do so, add the following lines to your `/config/configuration.yaml`
+without changing anything:
````
```suggestion
Since Home Assistant, by default, blocks requests from proxies/reverse proxies,
you need to tell your instance to allow requests from the Tailscale add-on.
In order to do so, add the following lines to your `/config/configuration.yaml`
without changing anything:
```

frenck commented on tailscale/DOCS.md:
````diff
@@ -17,8 +17,55 @@ the following URL:
+    - 127.0.0.1
+```
+
+**Note**: _There is no need to adapt anything in these lines since the addon
````
This is already mentioned in the section above. Let's remove it.

frenck commented on tailscale/DOCS.md:
```diff
@@ -17,8 +17,55 @@ the following URL:
+<b>(click to open)</b></i>
+</summary>
+
+### Home Assistant configuration for Tailscale Proxy feature
```
Let's move this section after the installation procedure

frenck commented on tailscale/DOCS.md:
```diff
@@ -17,8 +17,55 @@ the following URL:
+**Note**: _There is no need to adapt anything in these lines since the addon
+runs on your host network._
+
+### Tailscale configuration for Tailscale Proxy feature
+
+[DNS page][tailscale_dns]: Choose a **Tailnet name** and click **Enable HTTPS**
+under HTTPS Certificates (see [Enabling HTTPS][tailscale_info_https] for more
+information)
```
This should go after the installation, it is not a pre-requisite.

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```diff
@@ -44,22 +44,70 @@ do
   sleep 2
 done
+
+bashio::log.info "Tailscale is running"
+bashio::log.info "Tailscale version: $(/opt/tailscale version --daemon=false | head -n 1)"
```
This is already logged

```suggestion
```

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```diff
@@ -44,22 +44,70 @@ do
+
+bashio::log.info "Enabling Tailscale Proxy..."
+
+declare domain
```
Please declare at the start of the file.

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
Let's extract the proxy stuff out of this file, into its own single shot service

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```diff
@@ -44,22 +44,70 @@ do
+declare domain
+
+# Remove previously configured proxy first, because the configuration is persistent between reboots
+/opt/tailscale serve https:443 / "http://localhost:$(bashio::core.port)" off &> /dev/null || true
```
IMHO, we should check the current state and make it match and not just turn it off.

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
Instead, we should do nothing.
If it is turned off, we can just log that. All the rest is not needed.

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```suggestion
  bashio::log.warning "is not accessible through plain HTTP connection."
```

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```suggestion
  bashio::log.warning "If you want to use HTTPS certificates and HTTPS proxy on this device,"
```

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/post-tailscaled/run:
```suggestion
```

lmagyar commented:
I've rebased and rewritten as requested.

Also tested manually the proxy startup logic and some corner cases on real devices.

MelleD commented:
@lmagyar I'm looking forward to the PR 👍

frenck commented:
@MelleD Not sure, right now, the Home Assistant Core release has my priority this week.

lmagyar commented:
> Do you think it's possible to do a (beta) release this week?
> I would/could test it over the long Easter weekend :)

@MelleD You can test it with installing my forked repo, more test is better: https://github.com/lmagyar/homeassistant-addon-tailscale

MelleD commented:
> > Do you think it's possible to do a (beta) release this week?
> > I would/could test it over the long Easter weekend :)
> 
> @MelleD You can test it with installing my forked repo, more test is better: https://github.com/lmagyar/homeassistant-addon-tailscale

Ok great, I'll try that 👍

MelleD commented:
Short feedback:
Works for me and enabled it. Thanks a bunch

frenck commented on tailscale/rootfs/etc/s6-overlay/s6-rc.d/proxy/run:
So the thing that keeps me from merging this right now, is this line.

It basically sets/dictates it will expose Home Assistant on this port. For most use-cases, this is fine I guess, but I could also think of allowing more customization in the future. For example, use it to expose a completely different service (or additional custom services, for that matter).

I do see that out of the box, this is the wished-for situation though.

🤔

lmagyar replied:
Hmmm, something like this? This seems brutal. :) And not supported by the config options schema. :(
```
customize_proxy:
  active: true
  servers:
    - https_port: 443
      funnel: true
      locations:
        - mount_point: /
          local_port: 8123
        - mount_point: /foo/
          local_port: 8124
    - https_port: 8443
      funnel: false
      locations:
        - mount_point: /
          local_port: 8123
        - mount_point: /foo/
          local_port: 8124
```
- If active=false, it serves HA on 443 by default and turns on Funnel.
- Funnel is limited to 443, 8443 and 10000 by Tailscale, we should check it.
- Tailscale serve has memory, so on add-on start it would be better to delete any previous config and recreate it.
- Should I add Funnel functionality to this PR because of the common config?

frenck replied:
@Xe I know, I spoke with Shayne already.

Nevertheless, IMHO this is the wrong place to discuss such things and I would appreciate it if you'd not try to sidetrack the subjects of PR reviews for external gains.

Thanks 👍 

../Frenck

frenck replied:
@lmagyar Considering the future intentions/changes coming to Tailscale (via their web interface), I think we can just go this route. Once their UI rolls out, we can drop this part.

This keeps us from going into weird YAML configurations as this point.

WDYT?

lmagyar replied:
Yes I agree. I have no better idea even after sleeping a lot on this. :)

lmagyar commented:
FYI:
- I've squashed and rebased the PR.
- Added a taildrop like optional `bool?` configuration option to disable it in case somebody has any issues with it.

The "how to make it customizable" [review](https://github.com/hassio-addons/addon-tailscale/pull/137#pullrequestreview-1386515841) is still relevant/open.

I also plan to PR the Funnel feature after this PR settled.

frenck commented on tailscale/translations/en.yaml:
```diff
@@ -29,3 +29,9 @@ configuration:
+    name: Tailscale Proxy
+    description: >-
+      This option allows you to enable Tailscale's Proxy feature to present your
+      Home Assistant device on your tailnet with a valid certificate.
```
```suggestion
      Home Assistant instance on your tailnet with a valid certificate.
```

frenck commented on tailscale/DOCS.md:
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
+
+When not set, this option is enabled by default.
+
+Tailscale can provide a TLS certificate for your Home Assistant device within
```
```suggestion
Tailscale can provide a TLS certificate for your Home Assistant instance within
```

frenck commented on tailscale/DOCS.md:
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
+Tailscale can provide a TLS certificate for your Home Assistant device within
+your tailnet domain.
+
+This can prevent browsers to warn that HTTP URLs to your Home Assistant device
+look unencrypted (browsers are not aware of that connections between Tailscale
+nodes are secured with end-to-end encryption).
```
```suggestion
This can prevent browsers from warning that HTTP URLs to your Home Assistant device
look unencrypted (browsers are not aware of the connections between Tailscale
nodes are secured with end-to-end encryption).
```

frenck commented on tailscale/DOCS.md:
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
+
+More information: [Enabling HTTPS][tailscale_info_https]
+
+1. Configure Home Assistant to be accessible through HTTP connection (this is
```
```suggestion
1. Configure Home Assistant to be accessible through an HTTP connection (this is
```

frenck commented on tailscale/DOCS.md:
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
+   information. If you still want to use another HTTPS connection to access Home
+   Assistant, please use a reverse proxy add-on.
+
+1. Home Assistant, by default, blocks requests from reverse proxies, like the
+   Tailscale Proxy. In order to enable it, add the following lines to your
+   `configuration.yaml`, without changing anything:
```
```suggestion
1. Home Assistant, by default, blocks requests from reverse proxies, like the
   Tailscale Proxy. To enable it, add the following lines to your
   `configuration.yaml`, without changing anything:
```

frenck commented on tailscale/DOCS.md:
```diff
@@ -121,6 +122,49 @@ When not set, this option is enabled by default.
+
+1. Restart the add-on.
+
+**Note:** _You should not use any port number in the url that you used
```
```suggestion
**Note:** _You should not use any port number in the URL that you used
```

lmagyar commented:
FYI:
- I've accepted everything you suggested
- I've found another "device" -> "instance" replacement in DOCS.md
- Rebased to the latest main branch

frenck reviewed (approved):
Thanks, @lmagyar 👍 

../Frenck
